Add cil

paskino opened the pull request:
Adds [CIL optimisation](https://github.com/vais-ral/CCPi-Framework) and [regularisation](https://github.com/vais-ral/CCPi-Regularisation-Toolkit). 

Adds new iterative reconstruction algorithms from CIL. 
Currently tested with [SPDHG](https://github.com/CCPPETMR/SIRF/blob/spdhg_from_cil/examples/Python/PET/interactive/spdhg_reconstruction.py). 

Functionality requires: 

* https://github.com/CCPPETMR/SIRF/pull/237
* https://github.com/CCPPETMR/SIRF/pull/238
* Cython

paskino commented on SuperBuild/External_CCPi-RGL.cmake:
```diff
@@ -0,0 +1,153 @@
+      set(PYTHON_DEST "${CMAKE_INSTALL_PREFIX}/python")
+    endif()
+    message(STATUS "Python libraries found")
+    message(STATUS "SIRF Python modules will be installed in " ${PYTHON_DEST})
```
What do you refer to?

KrisThielemans replied:
`SIRF Python module` -> `CIL`?

paskino commented on SuperBuild/External_CCPi-Framework.cmake:
```diff
@@ -0,0 +1,127 @@
+  set(CMAKE_LIBRARY_PATH ${CMAKE_LIBRARY_PATH} ${SUPERBUILD_INSTALL_DIR})
+  set(CMAKE_INCLUDE_PATH ${CMAKE_INCLUDE_PATH} ${SUPERBUILD_INSTALL_DIR})
+
+  set(ENV{CIL_VERSION} 0.10.1)
```
The environment variable [is going to be removed](https://github.com/vais-ral/CCPi-Framework/pull/167) from the requirements for the build of CCPi-Framework.

paskino replied:
moved the variable to version_config as suggested

KrisThielemans commented:
a note on copyright dates and notices. If you copied code from somewhere, you need to preserve its copyright data/notice. At the same time, you need to make sure it's up to date with your changes.

Seems therefore that you either remove Ben and me, or you keep the original date. sorry

paskino commented on CMake/FindCython.cmake:
```diff
@@ -0,0 +1,49 @@
+# code adapted from 
+# https://github.com/symengine/symengine.py/blob/master/cmake/FindCython.cmake
```
https://github.com/CCPPETMR/SIRF-SuperBuild/blob/add_cil/CMake/FindCython.cmake#L3

KrisThielemans commented on version_config.cmake:
```diff
@@ -132,17 +135,44 @@ if (DEVEL_BUILD)
   set(DEFAULT_STIR_URL https://github.com/UCL/STIR )
   set(DEFAULT_STIR_TAG fd3a7576a11930856d6af50d217f17d4848c2bff)
 
+  ## Gadgetron
```
this need to be removed. Gadgetron and ISMRMRD should not be changed by this PR

KrisThielemans commented on .travis.yml:
```diff
@@ -27,6 +27,18 @@ cache:  # shared between builds
 #       i.e. -boost == -DUSE_SYSTEM_Boost=OFF, which means that Boost will be built.
 matrix:
  include:
+ - os: linux
```
these changes need to be removed.

casperdcl replied:
outdated comment?

paskino commented on SuperBuild/External_CCPi-Framework.cmake:
```diff
@@ -0,0 +1,124 @@
+    
+    CONFIGURE_COMMAND ""
+    BUILD_COMMAND ""
+    INSTALL_COMMAND ${CMAKE_COMMAND} -E env CIL_VERSION=0.10.0 SRC_DIR=${${proj}_BINARY_DIR} RECIPE_DIR=${${proj}_SOURCE_DIR}/Wrappers/Python/conda-recipe PYTHON=${PYTHON_EXECUTABLE} bash ${${proj}_SOURCE_DIR}/Wrappers/Python/conda-recipe/build.sh
```
currently pointing to v1.1.1

casperdcl commented:
Now getting a weird error during `make`:

```
Traceback (most recent call last):
  File "/home/travis/build/CCPPETMR/SIRF-SuperBuild/builds/CCPi-RGL/build/Wrappers/Python/setup-regularisers.py", line 6, in <module>
    from Cython.Distutils import build_ext
ImportError: No module named Cython.Distutils
```

even though `cython` was installed and found

rijobro commented:
1. I would change the cmake variable `cython_bin` to `cython_EXECUTABLE` (similar to `SWIG_EXECUTABLE` and `PYTHON_EXECUTABLE`).

2. On OSX, I'm getting an openMP error:
```
/Users/rich/Documents/Code/SIRF-SuperBuild/Build/sources/CCPi-RGL/Core/regularisers_CPU/FGP_TV_core.h:25:10: fatal error: 
      'omp.h' file not found
#include "omp.h"
```

I haven't looked through the code, but presumably you need inspire yourself from `BUILD_STIR_WITH_OPEN_MP` (default to `OFF` for `APPLE`), which then needs to get passed through to `CIL`.

paskino commented:
I see the OSX problem, but I've never tested the regularisation on OSX. I guess I'll disable CIL for OSX.

KrisThielemans commented on SuperBuild/External_CCPi-Framework.cmake:
```diff
@@ -0,0 +1,135 @@
+  )
+  endif()
+
+  mark_as_superbuild(
```
we have this in various places, but here it definitely doesn't make sense (i.e. there's no such variable)

KrisThielemans commented on SuperBuild/External_CCPi-Framework.cmake:
```diff
@@ -0,0 +1,135 @@
+   else()
+      if(${USE_SYSTEM_${externalProjName}})
+        find_package(${proj} ${${externalProjName}_REQUIRED_VERSION} REQUIRED)
+        message("USING the system ${externalProjName}, set ${externalProjName}_DIR=${${externalProjName}_DIR}")
```
```suggestion
        message("USING the system ${externalProjName}")
```
I don't think the `_DIR` variable is used at all (no CMake config)

KrisThielemans commented on SuperBuild/External_CCPi-FrameworkPlugins.cmake:
```diff
@@ -0,0 +1,124 @@
+  )
+  endif()
+
+  mark_as_superbuild(
```
same comments as above

KrisThielemans commented on SuperBuild/External_CCPi-RGL.cmake:
```diff
@@ -0,0 +1,156 @@
+    INSTALL_DIR ${libcilreg_Install_Dir}
+   # apparently this is the only way to pass environment variables to 
+   # external projects 
+    CONFIGURE_COMMAND ${CMAKE_COMMAND} -E env CIL_VERSION=${CIL_VERSION}         cmake ${${proj}_SOURCE_DIR}  -DCMAKE_INSTALL_PREFIX=${libcilreg_Install_Dir}
```
This will fail on Windows surely. Just add a comment

KrisThielemans commented on SuperBuild/External_CCPi-RGL.cmake:
```diff
@@ -0,0 +1,156 @@
+  )
+  endif()
+
+  mark_as_superbuild(
```
as above

KrisThielemans commented on docker/user_sirf-ubuntu.sh:
```diff
@@ -3,7 +3,7 @@
 INSTALL_DIR="${1:-/opt}"
 # SIRF
-git clone https://github.com/CCPPETMR/SIRF-SuperBuild --recursive -b master $INSTALL_DIR/SIRF-SuperBuild
+git clone https://github.com/CCPPETMR/SIRF-SuperBuild --recursive -b add_cil $INSTALL_DIR/SIRF-SuperBuild
```
will have to change

KrisThielemans replied:
don't forget!

KrisThielemans commented on version_config.cmake:
```diff
@@ -22,20 +22,23 @@
-     set(Boost_REQUIRED_VERSION 1.65.1)
-     set(Boost_URL http://downloads.sourceforge.net/project/boost/boost/${Boost_VERSION}/boost_1_65_1.zip)
-     set(Boost_MD5 9824a7a3e25c9d4fdf2def07bce8651c)
+    # Use version in Ubuntu 18.04
```
revert all these

KrisThielemans commented on version_config.cmake:
```diff
@@ -132,9 +135,18 @@ if (DEVEL_BUILD)
 
+else()
+	#  set(DEFAULT_SIRF_TAG v1.1.1)
+	set (DEFAULT_SIRF_TAG origin/add_to_sirf_classes)
```
will need to change

paskino commented:
I believe I addressed the requested changes. 

Travis fails to build during test because some packages aren't found (specifically python PIL, which is installed). I'm unsure if and how to install with pip, help is welcome.

I created a `BUILD_CIL` and `BUILD_CIL_LITE` options. With the latter one installs[ `CCPi-Framework`](https://github.com/vais-ral/CCPi-Framework), [`CCPi-FrameworkPlugins`](https://github.com/vais-ral/CCPi-FrameworkPlugins), [`CCPi-Regularisation-Toolkit`](https://github.com/vais-ral/CCPi-Regularisation-Toolkit). With the former one installs the additional X-ray CT engine [`ASTRA-toolbox`](https://github.com/astra-toolbox), [`CCPi-Astra`](https://github.com/vais-ral/CCPi-Astra), [`TomoPhantom`](https://github.com/dkazanc/TomoPhantom).
Currently these option are only available on Linux.

The things that are still to be changed are the SIRF TAG and CIL TAGs.

paskino commented:
```
STIR.py", line 1173
    if not out is None:
                      ^
TabError: inconsistent use of tabs and spaces in indentation
```

KrisThielemans commented:
I'm afraid I cannot accept a PR that makes most Travis jobs fail... I know there's OSX problems but this breaks even Linux.

@casperdcl would you be able to help with the `pip` problem?

casperdcl commented:
@KrisThielemans ah this looks fun. Over a hundred commits and comments. Haven't been following. What's the `pip` problem?

paskino commented:
@casperdcl I need the Python Image Library, PIL or pillow installed on travis.
it's installed but it's not found during tests. That's just the first build in the travis matrix. 

You can have a look (if you want) in the [log](https://travis-ci.org/CCPPETMR/SIRF-SuperBuild/jobs/589846076).

I guess the other failures are due to the previous comment.

casperdcl commented:
`PIL`'s deprecated in favour of the `Pillow` fork. You were also installing dependencies in the system rather than the working python env.

Should be fixed now (there are other unrelated errors).

casperdcl commented:
new problem:

 https://travis-ci.org/CCPPETMR/SIRF-SuperBuild/jobs/590322738#L2087

`TomoPhantom` is finding python2.7 rather than the env's `$PY_EXE`

paskino commented:
That's interesting. The way [TomoPhantom](https://github.com/dkazanc/TomoPhantom/blob/master/Wrappers/Python/CMakeLists.txt#L14) finds python is the same as [SIRF](https://github.com/CCPPETMR/SIRF/blob/master/CMakeLists.txt#L92)'s

casperdcl commented:
for the SuperBuild there's a `-DPYTHON_EXECUTABLE=$PY_EXE` which I assume is propagated to `SIRF`. You'll need the same for `TomoPhantom`

casperdcl commented:
sigh nvm just fixed it; you'll have to ensure `TomoPhantom` uses the env vars (https://github.com/CCPPETMR/SIRF-SuperBuild/pull/154/commits/432ff49cf65d16cc69e02e78b7f250f99bf9ba54) just as `SIRF` does.

[![](https://img.shields.io/badge/donate-casperdcl-ff69b4.svg?style=social&logo=Cash-App)](https://caspersci.uk.to/donate/)

paskino commented:
Thanks @casperdcl ! The last bit that fails is that travis builds the wrong branch of SIRF:

it should use [`add_to_sirf_classes`](
https://github.com/CCPPETMR/SIRF-SuperBuild/blob/432ff49cf65d16cc69e02e78b7f250f99bf9ba54/version_config.cmake#L186)

but in facts it uses `add_to_sirf_classes_merge_master`.

Does travis do some kind of caching?

casperdcl commented:
before your comment edit, about SIRF tags, it's probably best to specify `-DSIRF_TAG=...` somewhere in `.travis.yml` amd revert before merge rather than mess with `version_config`

paskino commented:
I see your point. I added an [`if (CIL_BUILD)`](https://github.com/CCPPETMR/SIRF-SuperBuild/blob/add_cil/version_config.cmake#L185) in `version_config` that will allow when `BUILD_CIL=ON` to build the `add_to_sirf_classes` branch of SIRF rather than `master` or `v2.0.0` 

There remain a few problems:
1. missing `wget` python module during tests (which I think I fixed)
2. osx [build](https://travis-ci.org/CCPPETMR/SIRF-SuperBuild/jobs/592466194#L16626)s
3. docker [build](https://travis-ci.org/CCPPETMR/SIRF-SuperBuild/jobs/592466200#L12276)

paskino commented:
@casperdcl docker builds fail because of some further dependencies [not met](https://travis-ci.org/CCPPETMR/SIRF-SuperBuild/jobs/592485766#L5727) for astra

casperdcl commented on CMake/FindCython.cmake:
```diff
@@ -0,0 +1,50 @@
+
+# This finds the "cython" executable in your PATH, and then in some standard
+# paths:
+SET(CYTHON_EXECUTABLE cython CACHE STRING "Cython executable name")
```
should we also be trying `python -m cython`?

KrisThielemans replied:
no idea really, but let's leave it. this seems based on other people's work

casperdcl commented on SuperBuild.cmake:
```diff
@@ -208,6 +209,21 @@ if (BUILD_petmr_rd_tools)
 
+if ("${PYTHON_STRATEGY}" STREQUAL "CONDA")
+  set (BUILD_CIL OFF)
+endif()
```
Why is `BUILD_CIL` overridden for `conda`?

paskino replied:
because it's not developed.

casperdcl replied:
what does that mean?

paskino replied:
it means that I developed only the installation for `PYTHONPATH`.

casperdcl replied:
right... not sure if that will work then. Surely the VM uses conda?

paskino replied:
No the VM uses system python and `PYTHONPATH`. I think Travis and docker use conda.

casperdcl replied:
kewl but as per https://github.com/CCPPETMR/SIRF-SuperBuild/blob/7354e5c9feafa46314df02be7ea508d472c4a56a/SuperBuild.cmake#L113-L116 there's no harm in using `CONDA`. It just means that whoever runs the build will be responsible for manually installing python modules themselves.

paskino replied:
In this case I don't think it should be there as an option.

casperdcl commented on SuperBuild/External_astra-toolbox.cmake:
```diff
@@ -0,0 +1,241 @@
+  message("astra-toolkit TAG " ${${proj}_TAG}  ) 
+
+  # conda build should never get here
+  if("${PYTHON_STRATEGY}" STREQUAL "PYTHONPATH")
```
where's the `endif()`?

paskino replied:
way down at line [213](https://github.com/CCPPETMR/SIRF-SuperBuild/blob/f30be5800259cafb3bc257d2daea9b16d89523be/SuperBuild/External_astra-toolbox.cmake#L213)

casperdcl replied:
indentation needs to be sorted out then

casperdcl commented on version_config.cmake:
```diff
@@ -159,10 +170,23 @@ if (DEVEL_BUILD)
+    set(DEFAULT_SIRF_TAG v2.0.0)
+  else()
+    set (DEFAULT_SIRF_TAG origin/add_to_sirf_classes)
+  endif()
```
will have to revert

paskino replied:
your suggestion is that building this branch will require manually setting `-DSIRF_TAG=origin/add_to_sirf_classes -DCCPi-Framework_TAG=...`

casperdcl replied:
yes I presume those branches will be merged at some point and v2.0.0 will be updated to v2.1.0 etc

casperdcl commented on version_config.cmake:
```diff
@@ -181,6 +205,19 @@ else()
+  
+  # CCPi CIL
+  set(CIL_VERSION "v19.07")
+  set(Regularisation-Toolkit_VERSION "19.06")
```
should have `DEFAULT_` prefix

KrisThielemans replied:
I think this can be ignored. These are internal variables that are used in the `DEFAULT` stuff below

casperdcl commented:
I'll be fixing the `docker/` stuff shortly (I hope).

casperdcl commented:
interesting looks like `travis` has an old version of `docker` which is failing to build because of something like https://github.com/GoogleContainerTools/skaffold/issues/2576 - or perhaps the cache was cleared in a previous build which failed.

KrisThielemans commented:
@casperdcl ready to merge?

casperdcl commented:
@KrisThielemans there's still https://github.com/CCPPETMR/SIRF-SuperBuild/pull/154/files#r330539512 I think needs to be addressed (or ignored - idk)

casperdcl commented:
apart from that there's all the hardcoded version/branch stuff which may have to change.

KrisThielemans commented on SuperBuild/External_NIFTYREG.cmake:
```diff
@@ -65,7 +65,8 @@ if(NOT ( DEFINED "USE_SYSTEM_${externalProjName}" AND "${USE_SYSTEM_${externalPr
       -DUSE_THROW_EXCEP=ON
       # fixes lib_reg_maths.a `GOMP_parallel' undefined reference linker errors
-      -DUSE_OPENMP:BOOL=OFF
+      -DUSE_OPENMP:BOOL=ON
```
I guess this will break Apple rather dramatically. @rijobro ?

In any case, we need `${proj}_ENABLE_OPENMP` as in https://github.com/CCPPETMR/SIRF-SuperBuild/blob/7a479cfd0b936da129e638932ca8b039f714791e/SuperBuild/External_STIR.cmake#L55-L60. To clean that up, we could have a `ENABLE_OPENMP` option in `SuperBuild.cmake`, and use its value in the various projects. But feel free to do that with a latter PR (in which case you'd need to raise an issue).

paskino replied:
opened https://github.com/CCPPETMR/SIRF-SuperBuild/issues/269

KrisThielemans commented on SuperBuild/External_TomoPhantom.cmake:
```diff
@@ -0,0 +1,108 @@
+# Author: Edoardo Pasca
+# Author: Casper da Costa-Luis
+# Copyright 2017, 2018 University College London
+# Copyright 2017, 2018 STFC
```
2019 as well?

KrisThielemans commented on version_config.cmake:
```diff
@@ -22,6 +22,8 @@
 #=========================================================================
 
 ## BOOST
+  # Gadgetron needs 1.65
```
```suggestion
  # Gadgetron needs at least 1.65
```

KrisThielemans commented:
@paskino, @casperdcl please add some lines to `CHANGES.md` as well before the merge please

casperdcl commented on SuperBuild/External_astra-toolbox.cmake:
```diff
@@ -0,0 +1,232 @@
+      PYTHONPATH: prefix PYTHONPATH \n\
+      SETUP_PY:   execute ${PYTHON_EXECUTABLE} setup.py install \n\
+      CONDA:      do nothing")
+    set_property(CACHE PYTHON_STRATEGY PROPERTY STRINGS PYTHONPATH SETUP_PY CONDA)
```
seems unneeded

casperdcl commented on SuperBuild/External_CCPi-Regularisation-Toolkit.cmake:
```diff
@@ -0,0 +1,159 @@
+      PYTHONPATH: prefix PYTHONPATH \n\
+      SETUP_PY:   execute ${PYTHON_EXECUTABLE} setup.py install \n\
+      CONDA:      do nothing")
+    set_property(CACHE PYTHON_STRATEGY PROPERTY STRINGS PYTHONPATH SETUP_PY CONDA)
```
seems unneeded

casperdcl commented:
@paskino looks like you squash-merged this which means no git history. @KrisThielemans would you mind if I force-push to master including the history?

casperdcl commented on .travis.yml:
```diff
@@ -207,6 +225,17 @@ before_install:
+- |
+   if [[ "$EXTRA_BUILD_FLAGS" == *"-DBUILD_CIL_LITE=ON"* ]]; then
+     $PY_EXE -m pip install --user -U scipy h5py Pillow wget
+   fi
```
inefficient - I'll fix this

casperdcl replied:
fixed in e8df19dc829b86617d17912a96a2be2e0ec33eb7

casperdcl commented on SuperBuild.cmake:
```diff
@@ -214,6 +216,25 @@ if (BUILD_petmr_rd_tools)
 endif()
 
+if ("${PYTHON_STRATEGY}" STREQUAL "CONDA")
+  set (BUILD_CIL OFF)
```
should also `set(BUILD_CIL_LITE OFF)`?

casperdcl commented on SuperBuild.cmake:
```diff
@@ -167,6 +167,8 @@ option(BUILD_STIR "Build STIR" ON)
 option(BUILD_siemens_to_ismrmrd "Build siemens_to_ismrmrd" OFF)
 option(BUILD_petmr_rd_tools "Build petmr_rd_tools" OFF)
+option(BUILD_CIL "Build CCPi CIL Modules and ASTRA engine" OFF)
+option(BUILD_CIL_LITE "Build CCPi CIL Modules" OFF)
```
better to have `BUILD_CIL` and `BUILD_ASTRA` separately? Or `BUILD_CIL_STRATEGY CACHE STRING "FULL NO_ASTRA"`?

KrisThielemans commented:
Ok. Please send an email to the list

Kris

________________________________
From: Casper da Costa-Luis <notifications@github.com>
Sent: Tuesday, October 8, 2019 9:58:37 AM
To: CCPPETMR/SIRF-SuperBuild <SIRF-SuperBuild@noreply.github.com>
Cc: Thielemans, Kris <k.thielemans@ucl.ac.uk>; Mention <mention@noreply.github.com>
Subject: Re: [CCPPETMR/SIRF-SuperBuild] Add cil (#154)


@paskino<https://github.com/paskino> looks like you squash-merged this which means no git history. @KrisThielemans<https://github.com/KrisThielemans> would you mind if I force-push to master including the history?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/CCPPETMR/SIRF-SuperBuild/pull/154?email_source=notifications&email_token=ABQRIHMVPYA3F52WZUUXVX3QNRDT3A5CNFSM4GEZX2D2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEATONXI#issuecomment-539420381>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ABQRIHPNOFFSP7JUR3NFYFLQNRDT3ANCNFSM4GEZX2DQ>.

casperdcl commented:
@KrisThielemans there was a 15min gap between the merge and my force-push, and the code was in the same state so don't think it should be an issue (@paskino you may be the only one who has to `git checkout master && git fetch origin && git reset --hard origin/master`
